### ISSUE TITLE
feat: Support updating images tagged with githash ids

### ISSUE THREAD

**Issue by hawksight** [argoproj-labs/argocd-image-updater#15]
**Is your feature request related to a problem? Please describe.**
I'm really excited to see argocd-image-updater repository as this feature is something I really felt I was missing our compared with flux and others like keel.sh. I've recently started using & migrating our application delivery process over to argocd and have really enjoyed seeing the development teams actually love the tool.

We have tagged images with the githash for a long time, and im sure many others have and will continue to do so. It's using the version tagging built into our version control tool of choice.. git. This is basically the only blocker to me trying out this tool.

Our current workflow is for devs to PR their new image githash, after mering their code the the respective branch. I want to make this process better.

**Describe the solution you'd like**
I would like to be able monitor for new version of my image tagged with a githash and auto deploy them, in the same way that this tool enables SemVer images to be deployed.

**Describe alternatives you've considered**

- keel.sh -> again this is on SemVer so can't got that way
- flux -> I don't want to use two deployment tools, its just twice the maintenance for me
- Converting 40+ applications to use a release process means changing our CI pipelines for all those applications, and fundamentally the company culture to embrace a new process. This is harder and longer process than implementing a tool to do it.

**Additional context**

I've dealt with this a little bit and will give some context to what I'm currently doing.

We use [App of Apps](https://argoproj.github.io/argo-cd/operator-manual/cluster-bootstrapping/#app-of-apps-pattern) for better or worse to manage all the applications in our environment. (Environment maps to a cluster which maps to a project in argocd) We are also using Declarative config from a single repository to define our ArgoCD application in yaml. Hence I have a tool (ansible) to generate the application config files, and also to regenerate them as needed.

One of the handy feature for me, it for ansible to go look at the applications git repository on a specified branch and return me the latest githash on that branch. That then feeds into my ArgoCD application yaml file that's generated. PR that to the ops repo and then the app of apps will sync it out.

I'd like in some way to replicate that back into the image-updater.

So perhaps to help with version constraints of githash'd images, we could look at using a git tag as a reference point.
For example maybe:

```yaml
argocd-image-updater.argoproj.io/image-list: <image>:git+<git-ref>
# could be:
argocd-image-updater.argoproj.io/image-list: <image>:git+staging
# compared with the SemVer:
argocd-image-updater.argoproj.io/image-list: nginx:~1.26
```

The idea being:
1) New image is available in container registry.
2) Argocd picks up the new change (webhook / poll) and stores `new_hash`
3) Argocd checks the constraint, notices it's git, and picks up the `staging` reference
4) Argocd checks the application repository for the referenced branch. (This may not be defined, but perhaps an annotation with the git url could be added to make this check easier, the credentials would have to be available to argo to query said repo of course).
5) Pulls the commit sha history for branch... and checks for `new_hash` in the commit history.
6) If in there, deploys. If not... it doesn't.

A further check could compare the date of `new_hash` with the existing `hash` in the application. Update but warn if the image is older that the current hash.
This might be a nice pattern for allow updates based on git branches. Something we use but there are many other strategies.

A side benefit of having argocd talk the git, is that it could potentially know when a new commit is available and show that a deployment is incoming. I.e. the time between merge of code in the app's repo and the CI building & pushing the image to the container registry.

Sorry that's a lot of ideas that bring man challenges, so ...

**Basic request**

Stage 1 - Just deploy the latest githash tagged image when available, with respect to sync options (manual / auto).

Stage 2 - Think about how we can improve the experience with more safeguards, like restricting to a git branch.

**Comment by jannfis** [comment]
Thanks a lot for this awesome and very detailed feature request, and also the reasoning behind it. I totally agree that this is a must-have feature, and we're certainly planning to allow updating to immutable, non-semver tags. It'll be some effort until this is possible, tho.

I think a viable path to get quickly to your proposed stage 1 would be to allow updating to semver-compatible tags containing Git hashes as the pre-release identifer, such as `v1.0.0-<githash>` instead of just using `<githash>` as the sole tag. Would that be something that fits in your use-case scenarios?

A little technical background: The container registry v2 API doesn't return a date along with tags, and returns an alphabetically ordered list of tags without any further meta-information. It's not possible to get this list sorted by publish date, and this obviously won't work with Git hashes. In order to get the publishing date of any given tag, another API call has to be performed for each tag in the repository to get the tag's manifest which contains the publishing data. So obviously, sorting the tags by date is a quite expensive operation to perform. So we definitely need a cache of already inspected tags (treating each of them as immutable), and only retrieve the dates from the Docker API for not yet known tags.

So I guess a quick path to get your stage 1 would therefore be:

* Implement a (non-persistent) tag list cache
* Retrieve tag metadata along with each tag from the registry and store it in cache (if there's a cache miss)
* If multiple tags for allowed semver restriction (i.e. `~v1.0.0` constraint should match `~v1.0.0-<githash>` - need to verify) are allowed, use the one with the latest publishing date as target tag.

As a step 1.5, we could improve the version matching algorithm to also consider non-semver tag specifications, providing additional methods (i.e. regexp) for selecting possible target tag candidates.

**Comment by hawksight** [comment]
> I think a viable path to get quickly to your proposed stage 1 would be to allow updating to semver-compatible tags containing Git hashes as the pre-release identifer, such as v1.0.0-<githash> instead of just using <githash> as the sole tag. Would that be something that fits in your use-case scenarios?

I think that is a fair middle ground to getting something that I could use. For example, even If I faked a semver by just prefixing the githash with the same semver string that could work. That would be an easy enough change for me, at least on some repositories to try it out.

I did not think to look back at the docker registry api, so I will go and take a look and have a think about strategies with the available endpoints.

One last thing, is that maybe don't think of tags as immutable. Images can be tagged and re-tagged so easily with little constraint. An example of this is...

- If a CI system re-runs its jobs, it'll produce another image and tag it with the same githash. So the old image has the tag removed and the newer image get the githash. 

It might be better to build the cache with image digests rather than the actual tag value (githash or otherwise). Digests can't change (i don't think). But definitely in favour of the caching this metadata, as I think that's the only way we'll be able to start doing clever comparison of when each image was published etc.

**Comment by jannfis** [comment]
As a quick heads up: The feature to update to the latest built images - so i.e. `v1.0.0-githash` - will be included with the upcoming v0.3.0 release. Probably going to release it this evening or tomorrow morning (CEST).

Regarding the cache, I wish it would be that simple. Unfortunately, the image digest is part of the tag's meta data, which needs to be fetched for each tag separately, so it would make no sense to use the digest as the cache key. There are registries who implemented their `/v2/image_name/tags/list` endpoint to deliver more meta data with this first request, but i.e. the Docker Hub registry does not. 

I hope this will change in the future, but for now, I think we must either treat image tags as immutable or do not cache tag meta data at all. However, as the canonical image name stays the same for different images in your use-case, I think getting a new version of that image running in your container is just a matter of `imagePullPolicy: always` and restarting the pods that make use of this container image. On the other hand, I would expect an image built from (and tagged to) a given point in time (i.e. a Git commit hash) is the same on subsequent builds. 

So maybe it'd be an even better option to include a sequential build number as additional metadata in the tag name, so for example, `v1.0.0-githash+N`, where N is the number of the build. This is perfectly covered by the semver v2 spec too.

**Comment by jannfis** [comment]
I'd be very grateful if you can have a look at the [v0.3.0 ](https://github.com/argoproj-labs/argocd-image-updater/releases/tag/v0.3.0) and see if it (at least partially) addresses your use-case. I'd be more than happy to fully solve it together with your help.

**Comment by aistein** [comment]
@jannfis how/where do users specify semver restrictions?  In our in-house image-updater solution, we enabled regex-based filtering of which image-push events from the docker registry would actually cause an update.   To do this we added a custom "annotation" in the `info` section of the `Application` CRD - looks something like this:
```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: my-application
spec:
  info:
    - name: gitops.argoproj.com/tag-regex-map
      value: >-
        my-docker-registry/my-image-1:^v\d+\.\d+\.\d+-.*,
        my-docker-registry/my-image-2:^v7+\.\d+\.\d+-.*
...
```

**Comment by jannfis** [comment]
@aistein You also define the list of images along with their version constraint in an annotation to the `Application`resource, such as:

```yaml
argocd-image-updater.argoproj.io/image-list: foo/bar:~1.17.0, foo/baz:^2.20.0
```
Version constraints are specified as semver ranges. Additional options and update strategies can be specified using another set of annotations. More info in [documentation](https://argocd-image-updater.readthedocs.io/en/stable/configuration/images/)

**Comment by hawksight** [comment]
Hey, sorry for taking so long. I've got a question on having mutliple images to update int he same application. Here's my code:

```
argocd-image-updater.argoproj.io/image-list: asfalia=gcr.io/sopost-k8s/asfalia:~v0.1.0-0, celery=gcr.io/sopost-k8s/asfalia:~v0.1.0-0
argocd-image-updater.argoproj.io/asfalia.update-strategy: latest
argocd-image-updater.argoproj.io/asfalia.helm.image-name: image.repository
argocd-image-updater.argoproj.io/asfalia.helm.image-tag: image.tag
argocd-image-updater.argoproj.io/celery.update-strategy: latest
argocd-image-updater.argoproj.io/celery.helm.image-name: celery.image.repository
argocd-image-updater.argoproj.io/celery.helm.image-tag: celery.image.tag
```

My helm chart values are as such:

```
image:
  repository: gcr.io/sopost-k8s/asfalia
  tag: ""
  pullPolicy: Always
celery:
  # When enabled, ensure all values are as expected
  enabled: false
  image:
    repository: gcr.io/sopost-k8s/asfalia
    tag: ""
    pullPolicy: IfNotPresent
```

It seems to update my `asfalia` image, but it does not want to update the `celery` image tag. It does look like in the debug logs that it is considering the image, just not updating it.

```
time="2020-08-27T09:59:05Z" level=info msg="Successfully updated image 'gcr.io/sopost-k8s/asfalia:v0.1.0-e46d4ba4a7ff398a387b28722332bb821cbf7129' to 'gcr.io/sopost-k8s/asfalia:v0.1.0-b5941f2ec8cb05590f1160bbae1a7ab25f943381'" application=ops-eu-asfalia image_name=sopost-k8s/asfalia image_tag=v0.1.0-e46d4ba4a7ff398a387b28722332bb821cbf7129 registry=gcr.io
time="2020-08-27T09:59:05Z" level=info msg="Processing results: applications=2 images_considered=3 images_updated=1 errors=0"
```

I have only one other application with a single image to update (that one is working).

So i guess, does my annotation configuration look correct, In the argocd UI it would seem only the main app `asfalia` is being overridden whereas the `celery` fields are not. Would the fact its the same image cause issue here?

Celery not being overridden:
![Screenshot from 2020-08-27 11-22-18](https://user-images.githubusercontent.com/11051419/91430980-9c483c80-e857-11ea-8c5a-08def111272b.png)

asfalia is overridden:
![Screenshot from 2020-08-27 11-22-01](https://user-images.githubusercontent.com/11051419/91430981-9ce0d300-e857-11ea-984b-1c4a3bb52e0b.png)

**Comment by hawksight** [comment]
Please ignore my above comment, issue was resolved in #75 

With regards to the original request, I can say that this is done:

> Just deploy the latest githash tagged image when available, with respect to sync options (manual / auto).

For those wanting an example, here's one of mine:

```
argocd-image-updater.argoproj.io/image-list: track=gcr.io/sopost-k8s/order-tracker
argocd-image-updater.argoproj.io/track.update-strategy: latest
argocd-image-updater.argoproj.io/track.helm.image-name: image.repository
argocd-image-updater.argoproj.io/track.helm.image-tag: image.tag
argocd-image-updater.argoproj.io/track.tag-match: regexp:^[a-z0-9]{40}$
```

This uses a helm chart where my image name and tag are specified as:

```
image:
  repository: <some-repo>/<app-name>
  tag: <githash>
```

Image-updater finds all images matching that regex... which is for a 40 character git commit sha. 
And will apply the latest image available.
eg: `<repo>/<app>:<githash>`

@jannfis has updated the docs with the simplest example for anyone else interested in using githashes:
https://argocd-image-updater.readthedocs.io/en/latest/configuration/images/#always-deploy-the-latest-build

I think the feature set so far makes it possible to do more targeted deploys by appending or including branch names in the image tagging (if you want to deploy apps from specific git branches). It also seems to follow the argocd philosophy of take what's in git and make it so. Completely decoupled from CI.

So I think we could also say that step 2 is also complete:

> Think about how we can improve the experience with more safeguards, like restricting to a git branch.

The only minor issue to note:
 - If you use `latest` update strategy but have an image with two tags. Image updater will continue to update the deployment from one tag to the next, in a loop.

So it's probably best to use a regex pattern to restrict to deployments to your known tag format, otherwise all it takes is for someone to tag the same image with something else, and you get a bit of a loop in of revisions in argocd.

@jannfis - I think we can close this thanks to your recent releases :+1:
